cellarium/cas/preprocessing/validator.py

```diff
@@ -33,10 +38,33 @@ def validate(
     cas_feature_schema_set = set(cas_feature_schema_list)
     adata_feature_schema_set = set(adata_feature_schema_list)
 
-    if adata_feature_schema_list == cas_feature_schema_list:
+    incompatible_x_type: t.Optional[str] = None
+    if count_matrix_input == constants.CountMatrixInput.X:
+        if adata.X.dtype != np.float32:
+            incompatible_x_type = adata.X.dtype
+    else:
+        if adata.raw.X.dtype != np.float32:
+            incompatible_x_type = adata.raw.X.dtype
+
+    incompatible_total_mrna_umis_type: t.Optional[str] = None
+    if "total_mrna_umis" in adata.obs and adata.obs["total_mrna_umis"].dtype != np.float32:
```

Comment:
`total_mrna_umis` is calculated in `pre_sanitize_callback` which executes before the `sanitize` (or actually first thing in `sanitize`). I think we might never have `total_mrna_umis` in this step, or am I missing something?

Reply:
Okay, I wasn't sure if `total_mrna_umis` was something that might already be in an anndata object that the user has.  I added that check because I assumed a user might already have it in their anndata object because we're not checking for it in `validate`.  It looks like `sanitize` only gets called if `validate` raises a `DataValidationError`, so it's possible that `total_mrna_umis` won't get added.  Should we just always call `sanitize`?  If so, do we drop `validate` completely?

Reply:
Hmmm. You're right. The thing is that handling DataValidationError  [results in logging what change is applied to adata](https://github.com/cellarium-ai/cellarium-cas/blob/main/cellarium/cas/client.py#L180). Also, if adata passes the validation, then it will bypass time consuming part of reordering the whole matrix. I think we need to factor out `pre_sanitize_callback` so that it'd be a third step in preprocessing:
1. pre_sanitize_callback
2. validate
3. sanitize if needed

Or maybe, another solution is to create a function `preprocessing` somewhere out of client class (in preprocessing module). In this case it would presanitize, validate, then sanitize and return `extra_features`, `missing_features` which would be used for logging for client.

Reply:
I can try one of those approaches.  The only issue I foresee is that, if the anndata matrix is the wrong type, it needs to be converted to float32 before the code in the pre_sanitize callback.  If it's not, the total_mrna_umis column will be the wrong type.  I could add the type check and conversion into the pre_sanitize_callback, maybe?  Or have it be a separate check that's part of a preprocessing function?  I assume we'd still want to log to the user that the type had to be converted.

Reply:
I think adding type chec into pre_sanitize_callback might be a good think!

Reply:
Awesome, that's the approach I'll take then.  Thanks!